File: app/tools/sql_guard.py
from __future__ import annotations

import re


class SQLGuard:
    allowed_objects = {
        'v_refund_metrics_daily',
        'v_ticket_sla',
        'v_ticket_detail',
        'v_recent_releases',
    }

    forbidden_patterns = [
        r'\binsert\b',
        r'\bupdate\b',
        r'\bdelete\b',
        r'\bdrop\b',
        r'\balter\b',
        r'\btruncate\b',
        r'\bcreate\b',
        r'\bunion\b',
        r'\boutfile\b',
        r'\binformation_schema\b',
        r'\bmysql\b',
        r'\bsleep\s*\(',
        r'\bbenchmark\s*\(',
        r'--',
        r'/\*',
        r'\*/',
        r'#',
        r';',
    ]

    def validate(self, sql: str) -> dict:
        candidate = sql.strip()
        lowered = candidate.lower()
        if not lowered.startswith('select'):
            return {'passed': False, 'severity': 'error', 'message': 'SQL 只允许 SELECT'}
        for pattern in self.forbidden_patterns:
            if re.search(pattern, lowered):
                return {'passed': False, 'severity': 'error', 'message': f'SQL 含有禁用内容: {pattern}'}

        limit_match = re.search(r'\blimit\s+(\d+)\b', lowered)
        if not limit_match:
            return {'passed': False, 'severity': 'error', 'message': 'SQL 必须带 LIMIT'}
        if int(limit_match.group(1)) > 200:
            return {'passed': False, 'severity': 'error', 'message': 'LIMIT 最大为 200'}

        objects = self._extract_objects(candidate)
        if not objects:
            return {'passed': False, 'severity': 'error', 'message': 'SQL 未识别到合法查询对象'}
        disallowed = [item for item in objects if item not in self.allowed_objects]
        if disallowed:
            return {'passed': False, 'severity': 'error', 'message': f'SQL 访问了非白名单对象: {", ".join(disallowed)}'}
        return {'passed': True, 'severity': 'info', 'message': 'SQL 校验通过'}

    def _extract_objects(self, sql: str) -> set[str]:
        matches = re.findall(
            r'\b(?:from|join)\s+([`"\[]?[a-zA-Z_][\w$]*(?:[`"\]]?\.[`"\[]?[a-zA-Z_][\w$]*[`"\]]?)?)',
            sql,
            flags=re.IGNORECASE,
        )
        objects: set[str] = set()
        for raw in matches:
            cleaned = raw.replace('`', '').replace('"', '').replace('[', '').replace(']', '')
            objects.add(cleaned.split('.')[-1])
        return objects

File: internal/app/sql_guard.go
package app

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

type GuardResult struct {
	Passed   bool   `json:"passed"`
	Severity string `json:"severity"`
	Message  string `json:"message"`
}

type SQLGuard struct {
	AllowedObjects    map[string]struct{}
	ForbiddenPatterns []*regexp.Regexp
}

func NewSQLGuard() *SQLGuard {
	patterns := []string{
		`\binsert\b`,
		`\bupdate\b`,
		`\bdelete\b`,
		`\bdrop\b`,
		`\balter\b`,
		`\btruncate\b`,
		`\bcreate\b`,
		`\bunion\b`,
		`\boutfile\b`,
		`\binformation_schema\b`,
		`\bmysql\b`,
		`\bsleep\s*\(`,
		`\bbenchmark\s*\(`,
		`--`,
		`/\*`,
		`\*/`,
		`#`,
		`;`,
	}
	compiled := make([]*regexp.Regexp, 0, len(patterns))
	for _, pattern := range patterns {
		compiled = append(compiled, regexp.MustCompile(pattern))
	}
	return &SQLGuard{
		AllowedObjects: map[string]struct{}{
			"v_refund_metrics_daily": {},
			"v_ticket_sla":           {},
			"v_ticket_detail":        {},
			"v_recent_releases":      {},
		},
		ForbiddenPatterns: compiled,
	}
}

func (g *SQLGuard) Validate(sql string, maxLimit int) GuardResult {
	candidate := strings.TrimSpace(sql)
	lowered := strings.ToLower(candidate)
	if !strings.HasPrefix(lowered, "select") {
		return GuardResult{Passed: false, Severity: "error", Message: "SQL 只允许 SELECT"}
	}
	for _, pattern := range g.ForbiddenPatterns {
		if pattern.MatchString(lowered) {
			return GuardResult{Passed: false, Severity: "error", Message: fmt.Sprintf("SQL 含有禁用内容: %s", pattern.String())}
		}
	}
	limitPattern := regexp.MustCompile(`\blimit\s+(\d+)\b`)
	matches := limitPattern.FindStringSubmatch(lowered)
	if len(matches) < 2 {
		return GuardResult{Passed: false, Severity: "error", Message: "SQL 必须带 LIMIT"}
	}
	limit, _ := strconv.Atoi(matches[1])
	if limit > maxLimit {
		return GuardResult{Passed: false, Severity: "error", Message: fmt.Sprintf("LIMIT 最大为 %d", maxLimit)}
	}
	objects := extractSQLObjects(candidate)
	if len(objects) == 0 {
		return GuardResult{Passed: false, Severity: "error", Message: "SQL 未识别到合法查询对象"}
	}
	disallowed := make([]string, 0)
	for _, object := range objects {
		if _, ok := g.AllowedObjects[object]; !ok {
			disallowed = append(disallowed, object)
		}
	}
	if len(disallowed) > 0 {
		return GuardResult{Passed: false, Severity: "error", Message: fmt.Sprintf("SQL 访问了非白名单对象: %s", strings.Join(disallowed, ", "))}
	}
	return GuardResult{Passed: true, Severity: "info", Message: "SQL 校验通过"}
}

func extractSQLObjects(sql string) []string {
	pattern := regexp.MustCompile(`(?i)\b(?:from|join)\s+([` + "`" + `"\[]?[a-zA-Z_][\w$]*(?:[` + "`" + `"\]]?\.[` + "`" + `"\[]?[a-zA-Z_][\w$]*[` + "`" + `"\]]?)?)`)
	matches := pattern.FindAllStringSubmatch(sql, -1)
	seen := make(map[string]struct{})
	result := make([]string, 0, len(matches))
	for _, match := range matches {
		if len(match) < 2 {
			continue
		}
		cleaned := strings.NewReplacer("`", "", `"`, "", "[", "", "]", "").Replace(match[1])
		parts := strings.Split(cleaned, ".")
		name := parts[len(parts)-1]
		if _, ok := seen[name]; ok {
			continue
		}
		seen[name] = struct{}{}
		result = append(result, name)
	}
	return result
}
